cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -0,0 +1,363 @@
+    P^(meas, m)_|...01>, P^(meas, m)_|...10>, P^(meas, m)_|...11>... at the
+    end of circuit_mn for all m. This is done by taking a total of R
+    projective measurements for each circuit_mn, histogram the collected
+    bit-strings, and normalize the number of occurrences of each bit-string
+    by R. Bit-strings not seen in the measurements are assigned zero measured
+    probability and therefore do not contribute to the estimation of the
+    experimental cross-entropy s_meas (see below).
+
+    2) Theoretically compute the expected bit-string probabilities
+    P^(exp, m)_|...00>,  P^(exp, m)_|...01>, P^(exp, m)_|...10>,
+    P^(exp, m)_|...11>... at the end of circuit_mn for all m and for all
+    possible bit-strings of the Hilbert space.
+
+    3) Collect measured probabilities of all M circuits into a single vector
+    P_meas and normalize P_meas by M such it sums to 1.
+
+    4) Collect all expected probabilities into a single vector P_exp. Set any
+    probabilities <1e-22 to 1e-22 to avoid singularities in log P_exp. Then
+    normalize P_exp by M such it sums to 1.
+
+    5) Calculate the "XEB fidelity" at this given circuit depth using the
+    following expressions:
+
+    s_incoherent = -np.sum(P_uniform * np.log(P_exp))
+    s_expected = -np.sum(P_exp * np.log(P_exp))
+    s_meas = -np.sum(P_meas * np.log(P_exp))
+    fidelity = (s_incoherent - s_meas) / (s_incoherent - s_expected)
+
+    where P_uniform = 1/(M * 2 ** num_qubits) is a constant probability
+    expected from a uniform distribution for all M circuits, with num_qubits
+    being the total number of qubits.
+
+
+    Args:
+        sampler: The quantum engine or simulator to run the circuits.
+        qubits: The qubits included in the XEB experiment.
+        benchmark_ops: The set(s) of operations which are to be benchmarked for
+            fidelity. If more than one set is specified, the random circuits
+            will rotate between the sets. As an example, if benchmark_ops = [
+            {ops.CZ(q0, q1), ops.CZ(q2, q3)}, {ops.CZ(q1, q2)}] where q0, q1,
+            q2 and q3 are instances of Qid (such as GridQubits), each random
+            circuit will apply CZ gate between q0 and q1 plus CZ between q2
+            and q3 for the first cycle, CZ gate between q1 and q2 for the
+            second cycle, CZ between q0 and q1 and CZ between q2 and q3 for
+            the third cycle and so on. If None, the circuits will consist only
+            of single-qubit gates.
+        num_circuits: The total number of random circuits to be used.
+        repetitions: The number of measurements for each circuit to estimate
+            the bit-string probabilities.
+        cycles: The different numbers of circuit layers in the XEB study.
+            Could be a single or a collection of values.
+        single_qubit_gates: If None (by default), the single-qubit
+            gates are chosen from X/2 (\pi/2 rotation around the X axis),
```

Comment:
Use unicode character for pi, or use $ around latex.

Reply:
Done.